### DISCUSSION TITLE
intro to merkel trees

### DISCUSSION THREAD

**Discussion by jnorthrup** [xtclang/xvm#16] [Ideas]
if you haven't given any thought to the matter of snapshotting and idempotence on disk IPFS is a storage DAG done right.  [Merkel trees]( https://flyingzumwalt.gitbooks.io/decentralized-web-primer/content/ipfs-dag/ ) can deliver a pretty strong assurance for a DAG of subtree/hashes of a complete VFS.

for a VM/classloader/snapshotting notion adder48 or crc48 (adler/crc32+adler/crc16) can probably give you the cheapest usable baseline file hashes against which to strobe your mmap blocks and form a more significant point-in-time Merkle hash tree (like sha).  certainly no fort knox but if you wanted 512 bit hashes on ~4k chunks  you could pay for it there too for a given vertical.

**Comment by cpurdy** [comment]
That's an interesting idea.

For those who do not have context for Jim's suggestion, the background is that a question popped up on the Coffee Compiler Club about how Ecstasy could store off a container (e.g. to disk) when the container contains **stateful** injected references, such as a file in the process of being read.

The first part (a missing design item) that we need to figure out is how the Container API needs to change to allow the owner of the container to participate in the process of deciding how injected references are to be represented. For example, the owner of the container (a) may want to disallow this altogether, or (b) may want to be able to inject a poison reference when the container is restored, or (c) may want to be asked when the container is restored to provide a suitable injection at that point (which implies that it may need to store off additional information at _store_ time that it can then refer to at _restore_ time).
